# Weather Service IPM Interface

In [6]:
from agroservices.ipm import IPM
from agroservices.datadir import ipm_datadir
ipm = IPM()

## MetaDataServices

* Get IPM Platform weather parameters and their description (id, name, description and units)

In [2]:
ipm.get_parameter()

[{'id': 1001,
  'name': 'Instantaneous temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1002,
  'name': 'Mean air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1003,
  'name': 'Minimum air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MIN'},
 {'id': 1004,
  'name': 'Maximum air temperature at 2m',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MAX'},
 {'id': 1021,
  'name': 'Instantaneous temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1022,
  'name': 'Mean air temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'AVG'},
 {'id': 1023,
  'name': 'Minimum air temperature in canopy',
  'description': None,
  'unit': 'Celcius',
  'aggregationType': 'MIN'},
 {'id': 1024,
  'name': 'Maximum air temperature in canopy',
  'description': No

* Get the controle quality informations (id, name and description)

In [3]:
ipm.get_qc()

[{'id': 0,
  'name': 'No quality control performed',
  'description': 'There has been no quality control of the parameter'},
 {'id': 1,
  'name': 'OK. Parameter comes from a qualified external source',
  'description': 'The external source has its own quality controls, so we can trust it'},
 {'id': 2,
  'name': 'OK. Quality control performed by IPM Decisions',
  'description': "The parameter has been QC'd by the IPM Decisions platform and found to be OK"},
 {'id': 4,
  'name': 'Failed. Not an actual number',
  'description': 'The value is not a valid number'},
 {'id': 8, 'name': 'Failed. Interval test', 'description': 'TODO'},
 {'id': 16, 'name': 'Failed. Logical test', 'description': 'TODO'},
 {'id': 32, 'name': 'Failed. Step test', 'description': 'TODO'},
 {'id': 64, 'name': 'Failed. Freeze test', 'description': 'TODO'},
 {'id': 128,
  'name': 'Failed. Missing value(s)',
  'description': 'Either empty or NULL values have been detected'}]

* Get a schema that describes the IPM Decision platform's format for exchange of weather data

In [4]:
ipm.get_schema_weatherdata() 

{'$schema': 'http://json-schema.org/draft-04/schema#',
 'title': 'Weather Data',
 'type': 'object',
 'additionalProperties': False,
 'description': 'Version 0.1. The schema describes the weather data format for the IPM Decisions platform. See an example here: https://platform.ipmdecisions.net/api/wx/rest/weatheradapter/yr/?longitude=14.3711&latitude=67.2828&altitude=70',
 '$id': 'https://platform.ipmdecisions.net/api/wx/rest/schema/weatherdata',
 'properties': {'timeStart': {'type': 'string',
   'format': 'date-time',
   'description': 'The timestamp of the first weather observation. Format: "yyyy-MM-dd\'T\'HH:mm:ssXXX", e.g. 2020-04-09T18:00:00+02:00',
   'title': "Time start (yyyy-MM-dd'T'HH:mm:ssXXX)"},
  'timeEnd': {'type': 'string',
   'format': 'date-time',
   'description': 'The timestamp of the last weather observation. Format: "yyyy-MM-dd\'T\'HH:mm:ssXXX", e.g. 2020-04-09T18:00:00+02:00',
   'title': "Time end (yyyy-MM-dd'T'HH:mm:ssXXX)"},
  'interval': {'type': 'integer',
   

* Validates the posted weather data against the Json schema

In [7]:
ipm.post_schema_weatherdata_validate(jsonfile=ipm_datadir + 'weather_data.json')

{'isValid': True}

## Weather Adapter Services

* Get a list of WeatherAdapterService available on ipm plateform

In [11]:
ws = ipm.get_weatherdatasource()
{k:v['description'] for k,v in ws.items()}

{'no.met.locationforecast': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
 'ie.gov.data': '9-day forecasts for Ireland.',
 'dk.dmi.pointweather': 'Seasonal weather data and forecasts for Denmark. Grid based',
 'se.slu.lantmet': 'Seasonal weather data and forecasts for Sweden. Grid based',
 'net.ipmdecisions.dwd.forecast.germany': "27 hour weather forecasts for Germany and surrounding areas. 2.5 km resolution. Pulls data from DWD's Open Data server (https://www.dwd.de/EN/ourservices/opendata/opendata.html). Made available by a custom adapter created by Met Norway and NIBIO.",
 'net.ipmdecisions.dwd.forecast.europe': "30 hour weather forecasts for EU. 7 km resolution. Pulls data from DWD's Open Data server (https://www.dwd.de/EN/ourservices/opendata/opendata.html). Made available by a custom adapter created by Met Norway and NIBIO.",
 'net.ipmdecisions.dwd.euroweather': "Seasonal data and 72 hour weather forecasts for EU. 7 km reso

* Get weather observations for one weatheradapter service

In this example we choose Finnish Meteorological Institute measured data using endpoint='/weatheradapter/fmi/'

This weather resource does not need credentials, so credentials is None    
ignone= True    
interval=3600    
parameters = [1002, 3002] to select weather parameters 'Mean air temperature at 2m' and 'Mean RH at 2m' see ipm.get_parameters() function for description    
timeStart='2020-06-12T00:00:00+03:00'    
timeEnd='2020-07-03T00:00:00+03:00'    
weatherStationId=101104

In [12]:
source = ipm.get_weatherdatasource('fi.fmi.observation.station')
ipm.get_weatheradapter(source)

{'timeStart': '2020-01-01T23:00:00Z',
 'timeEnd': '2020-01-02T23:00:00Z',
 'interval': 3600,
 'weatherParameters': [2001, 1002, 3002],
 'locationWeatherData': [{'longitude': 24.26084,
   'latitude': 63.08898,
   'altitude': 0.0,
   'amalgamation': [0, 0, 0],
   'data': [[0.0, 4.8, 77.0],
    [0.0, 4.7, 75.0],
    [0.0, 4.3, 77.0],
    [0.0, 3.9, 79.0],
    [0.0, 3.7, 81.0],
    [0.0, 3.8, 80.0],
    [0.0, 3.6, 81.0],
    [0.0, 3.7, 81.0],
    [0.0, 4.2, 80.0],
    [0.1, 3.6, 89.0],
    [0.4, 3.5, 94.0],
    [0.3, 3.4, 96.0],
    [0.0, 3.4, 87.0],
    [0.0, 3.3, 84.0],
    [0.0, 3.1, 86.0],
    [0.0, 3.0, 85.0],
    [0.0, 3.1, 85.0],
    [0.0, 3.0, 86.0],
    [0.0, 3.2, 84.0],
    [0.0, 3.3, 84.0],
    [0.0, 3.3, 84.0],
    [0.0, 3.3, 85.0],
    [0.0, 3.2, 87.0],
    [0.0, 3.1, 88.0],
    [0.0, 3.5, 87.0]],
   'length': 25,
   'width': 3,
   'qc': [1, 1, 1]}]}

<div class="alert alert-warning">

 Among Weather resource forecast resources are available. When you choose forecast resource thank to use get_weatheradappter_forecast

<div>

In [15]:
source = ipm.get_weatherdatasource('no.met.locationforecast')
params = ipm.weatheradapter_forecast_params(source, altitude=70, latitude=67.2828, longitude=14.3711)
ipm.get_weatheradapter(source, params)


{'timeStart': '2023-05-15T15:00:00Z',
 'timeEnd': '2023-05-25T12:00:00Z',
 'interval': 3600,
 'weatherParameters': [1001, 3001, 2001, 4002],
 'locationWeatherData': [{'longitude': 14.3711,
   'latitude': 67.2828,
   'altitude': 70.0,
   'amalgamation': [0, 0, 0, 0],
   'data': [[11.4, 80.5, 0.0, 5.6],
    [10.7, 84.5, 0.0, 5.0],
    [10.5, 84.2, 0.0, 4.8],
    [10.4, 86.5, 0.0, 3.8],
    [10.2, 87.8, 0.0, 3.3],
    [9.8, 88.4, 0.0, 5.0],
    [9.3, 89.3, 0.0, 4.9],
    [8.7, 89.6, 0.0, 4.0],
    [8.5, 88.3, 0.0, 3.8],
    [8.3, 86.3, 0.0, 3.2],
    [8.3, 85.4, 0.0, 2.7],
    [8.0, 86.3, 0.0, 2.2],
    [8.0, 87.0, 0.0, 2.3],
    [8.1, 87.8, 0.0, 2.4],
    [8.1, 86.7, 0.0, 2.5],
    [8.3, 86.5, 0.0, 2.2],
    [8.4, 88.6, 0.0, 1.9],
    [8.4, 89.0, 0.0, 2.1],
    [8.3, 86.4, 0.3, 2.3],
    [8.4, 84.8, 0.2, 1.3],
    [8.3, 83.8, 0.1, 1.5],
    [8.5, 82.8, 0.1, 1.4],
    [8.5, 80.8, 0.2, 1.4],
    [8.5, 83.8, 0.2, 1.8],
    [8.2, 85.6, 0.2, 1.9],
    [8.0, 82.8, 0.3, 1.7],
    [7.8, 84.1, 0.

## WeatherDataService

* Get a list of all the available weather data sources

In [16]:
ipm.get_weatherdatasource().keys()

dict_keys(['no.met.locationforecast', 'ie.gov.data', 'dk.dmi.pointweather', 'se.slu.lantmet', 'net.ipmdecisions.dwd.forecast.germany', 'net.ipmdecisions.dwd.forecast.europe', 'net.ipmdecisions.dwd.euroweather', 'net.ipmdecisions.meteofrance.forecast', 'fi.fmi.forecast.location', 'fi.fmi.observation.station', 'no.nibio.lmt', 'com.meteobot', 'info.fruitweb', 'net.ipmdecisions.metos', 'fr.meteo-concept.api'])

* Search for weather data sources that serve the specific location. The location can by any valid Geometry, such as Point or Polygon. Example GeoJson input 

In [18]:
ipm.post_weatherdatasource_location(tolerance=0,geoJsonfile=ipm_datadir + 'GeoJson.json')

[{'id': 'no.met.locationforecast',
  'name': 'Met Norway Locationforecast',
  'description': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
  'public_URL': 'https://api.met.no/weatherapi/locationforecast/1.9/documentation',
  'endpoint': '{WEATHER_API_URL}/rest/weatheradapter/yr/',
  'authentication_type': 'NONE',
  'needs_data_control': 'false',
  'access_type': 'location',
  'priority': 0,
  'temporal': {'forecast': 9,
   'historic': {'start': None, 'end': None},
   'intervals': [3600]},
  'parameters': {'common': [1001, 3001, 2001, 4002], 'optional': None},
  'spatial': {'countries': None, 'geoJSON': '{"type": "Sphere"}'},
  'organization': {'name': 'The Norwegian Meteorological Institute',
   'country': 'Norway',
   'address': 'Postboks 43 Blindern',
   'postal_code': '0371',
   'city': 'Oslo',
   'email': 'david.melchior@met.no',
   'url': 'https://www.met.no/en'},
  'active': True,
  'endpointFullPath': 'https://platform.ipm

* Search for weather data sources that serve the specific point

In [19]:
ipm.get_weatherdatasource_location_point(
    latitude="59.678835236960765",
    longitude="12.01629638671875",
    tolerance=0)

[{'id': 'no.met.locationforecast',
  'name': 'Met Norway Locationforecast',
  'description': '9-day forecasts for the entire planet. 2.5 km resolution in the Nordic-Baltic region, 9km elsewhere',
  'public_URL': 'https://api.met.no/weatherapi/locationforecast/1.9/documentation',
  'endpoint': '{WEATHER_API_URL}/rest/weatheradapter/yr/',
  'authentication_type': 'NONE',
  'needs_data_control': 'false',
  'access_type': 'location',
  'priority': 0,
  'temporal': {'forecast': 9,
   'historic': {'start': None, 'end': None},
   'intervals': [3600]},
  'parameters': {'common': [1001, 3001, 2001, 4002], 'optional': None},
  'spatial': {'countries': None, 'geoJSON': '{"type": "Sphere"}'},
  'organization': {'name': 'The Norwegian Meteorological Institute',
   'country': 'Norway',
   'address': 'Postboks 43 Blindern',
   'postal_code': '0371',
   'city': 'Oslo',
   'email': 'david.melchior@met.no',
   'url': 'https://www.met.no/en'},
  'active': True,
  'endpointFullPath': 'https://platform.ipm